src/bin/pubsub_server.rs

```diff
@@ -0,0 +1,170 @@
+extern crate actix_web;
+#[macro_use]
+extern crate log;
+extern crate actix;
+extern crate env_logger;
+#[macro_use]
+extern crate serde_derive;
+extern crate clap;
+extern crate futures;
+extern crate pubsub;
+extern crate serde;
+
+use actix_web::State as ActixState;
+use actix_web::{
+    http::Method, middleware::Logger, server, App, AsyncResponder, Error, HttpMessage, HttpRequest,
+    HttpResponse, Path,
+};
+use clap::{App as ClApp, Arg};
+use futures::future::Future;
+use pubsub::{ConnectionInfo, PublisherDesc};
+use std::collections::HashMap;
+use std::net::ToSocketAddrs;
+use std::option::Option;
+use std::sync::{Arc, RwLock};
+use std::time;
+
+#[derive(Serialize, Debug)]
+struct BlankResponse {}
+
+#[derive(Serialize, Debug)]
+struct ConnectionResponse {
+    publisher: PublisherDesc,
+    info: ConnectionInfo,
+}
+
+#[derive(Serialize, Debug)]
+struct PubSubResponse<'a, T>
+where
+    T: serde::Serialize + 'a,
+{
+    status: &'a str,
+    timestamp: time::SystemTime,
+    response: &'a T,
+}
+
+impl<'a> PubSubResponse<'a, BlankResponse> {
+    fn status_only(status: &'a str) -> PubSubResponse<'a, BlankResponse> {
+        PubSubResponse {
+            status: status,
+            timestamp: time::SystemTime::now(),
+            response: &BlankResponse {},
+        }
+    }
+}
+
+impl<'a, T> PubSubResponse<'a, T>
+where
+    T: serde::Serialize,
+{
+    fn new(status: &'a str, response: &'a T) -> PubSubResponse<'a, T> {
+        PubSubResponse {
+            status: status,
+            timestamp: time::SystemTime::now(),
+            response: response,
+        }
+    }
+}
+#[derive(Serialize, Debug)]
+struct ServiceStatus {
+    count: usize,
+}
+
+type State = HashMap<String, ConnectionResponse>;
+type ProtectedState = Arc<RwLock<State>>;
+
+fn handle_service_status_request(req: HttpRequest<ProtectedState>) -> HttpResponse {
+    let state = req.state().read().unwrap();
+    let status = ServiceStatus { count: state.len() };
+    HttpResponse::Ok().json(PubSubResponse::new("Success", &status))
+}
+
+fn handle_add_connection(
+    req: HttpRequest<ProtectedState>,
+) -> Box<Future<Item = HttpResponse, Error = Error>> {
+    let protected_state = Arc::clone(req.state());
+    req.json()
+        .from_err()
+        .and_then(move |publisher: PublisherDesc| {
+            let name = publisher.name.clone();
+            let info = ConnectionResponse {
+                publisher: publisher,
```

Comment:
Name name name name

Reply:
I was just thinking that the connection info doesn't need the name. it was mostly there to allow those to be separated. Now that I am into this, I don't think the duplication is needed.